src/pybind/mgr/rook/rook_cluster.py

```diff
@@ -473,15 +472,165 @@ def filter_devices(self, rook_pods: KubernetesResource, drive_group: DriveGroupS
+                raise RuntimeError(err)
+
+    def set_osds_down(self) -> None:
+        down_flag_args = {
+            'prefix': 'osd down',
+            'ids': [str(x) for x in self.osd_ids]
+        }
+        ret, out, err = self.mon_command(down_flag_args)
+        if ret != 0:
+            log.error(f"error setting down flags on OSDs, {err}")
+            raise RuntimeError(err)
+
+    def scale_deployments(self) -> None:
+        for osd_id in self.osd_ids:
+            self.appsV1_api.patch_namespaced_deployment_scale(namespace='rook-ceph', name='rook-ceph-osd-{}'.format(osd_id), body=client.V1Scale(
+                spec=client.V1ScaleSpec(
+                    replicas=0
+                )
+            ))
+
+    def set_osds_out(self) -> None:
+        out_flag_args = {
+            'prefix': 'osd out',
+            'ids': [str(x) for x in self.osd_ids]
+        }
+        ret, out, err = self.mon_command(out_flag_args)
+        if ret != 0:
+            log.error(f"error setting down flags on OSDs, {err}")
+            raise RuntimeError(err)
+            
+    def delete_deployments(self) -> None:
+        for osd_id in self.osd_ids:
+            self.appsV1_api.delete_namespaced_deployment(namespace='rook-ceph', name='rook-ceph-osd-{}'.format(osd_id), propagation_policy='Foreground')
+
+    def clean_up_prepare_jobs_and_pvc(self) -> None:
+        for job in self.jobs.items:
+            if job.metadata.labels['ceph.rook.io/pvc'] in self.pvc_to_remove:
+                self.batchV1_api.delete_namespaced_job(name=job.metadata.name, namespace='rook-ceph', propagation_policy='Foreground')
+                self.coreV1_api.delete_namespaced_persistent_volume_claim(name=job.metadata.labels['ceph.rook.io/pvc'], namespace='rook-ceph', propagation_policy='Foreground')
+
+    def purge_osds(self) -> None:
+        for id in self.osd_ids:
+            purge_args = {
+                'prefix': 'osd purge-actual',
+                'id': int(id),
+                'yes_i_really_mean_it': True
+            }
+            ret, out, err = self.mon_command(purge_args)
+            if ret != 0:
+                log.error(f"error setting down flags on OSDs, {err}")
+                raise RuntimeError(err)
+
+    def destroy_osds(self) -> None:
+        for id in self.osd_ids:
+            destroy_args = {
+                'prefix': 'osd destroy-actual',
+                'id': int(id),
+                'yes_i_really_mean_it': True
+            }
+            ret, out, err = self.mon_command(destroy_args)
+            if ret != 0:
+                log.error(f"error setting down flags on OSDs, {err}")
+                raise RuntimeError(err)
+
+    def remove(self) -> str:
+        self.check_force()
+        try:
+            remove_result = self.remove_device_sets()
+        except Exception as e:
+            raise RuntimeError("Failed to update CephCluster CR, removing OSDs failed: " + str(e))
```

Comment:
mmm.. with this change, if we fail patching the cluster CRD we are going to continue doing all the deletion of deployments, jobs, and setting OSDs down and out .. does not seem very healthy.
I suggest to "return if the cluster CRD patch cannot be done...

Another important thing, with the error management we have now, any error seems to be a programming error , with the nice dump of instructions,  that probably is going to scare  the final user ,, making him/her cry and run away from the terminal ...

Example: We try to delete an OSD, but that is not possible because is not safe ( the cluster needs the OSD, deleting it will cause problems or even lost of information) .  What we have now is:

```
[root@rook-ceph-tools-78cdfd976c-s4d6h /]# ceph orch osd rm 2
Error EINVAL: Traceback (most recent call last):
  File "/usr/share/ceph/mgr/mgr_module.py", line 1564, in _handle_command
    return self.handle_command(inbuf, cmd)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 167, in handle_command
    return dispatch[cmd['prefix']].call(self, cmd, inbuf)
  File "/usr/share/ceph/mgr/mgr_module.py", line 415, in call
    return self.func(mgr, **kwargs)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 107, in <lambda>
    wrapper_copy = lambda *l_args, **l_kwargs: wrapper(*l_args, **l_kwargs)  # noqa: E731
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 96, in wrapper
    return func(*args, **kwargs)
  File "/usr/share/ceph/mgr/orchestrator/module.py", line 829, in _osd_rm_start
    completion = self.remove_osds(osd_id, replace=replace, force=force)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 1360, in inner
    completion = self._oremote(method_name, args, kwargs)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 1430, in _oremote
    return mgr.remote(o, meth, *args, **kwargs)
  File "/usr/share/ceph/mgr/mgr_module.py", line 1951, in remote
    args, kwargs)
RuntimeError: Remote method threw exception: Traceback (most recent call last):
  File "/usr/share/ceph/mgr/rook/module.py", line 484, in remove_osds
    res = self._rook_cluster.remove_osds(osd_ids, replace, force, self.mon_command)
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 1031, in remove_osds
    return self.remover.remove()
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 609, in remove
    self.check_force()
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 544, in check_force
    raise RuntimeError(err)
RuntimeError: OSD(s) 2 have 128 pgs currently mapped to them. 
```


That is not a programming error, is an usage error. Usage errors must be reported in a friendly way and need to be informative enough to make easy for the user to know what is the better way to solve the problem.
In this cases, we only need to show the error message and not the stack trace. (the stack trace is very useful in the log)

So i suggest you to change the error management:
clues:
- In the methods called by "remove" if we have an exception just "Raise" the same error

- In the remove method.. i think that better to explain with code:
```python
    def remove(self) -> str:
        self.check_force()
        try:
            remove_result = self.remove_device_sets()
        except Exception as e:
            log.exception("Error patching ceph cluster CRD")
            return f"Not possible to modify Ceph cluster CRD: {e}"
        try:
            self.scale_deployments()
            self.delete_deployments()
            self.clean_up_prepare_jobs_and_pvc()
        except Exception as e:
            log.exception("Ceph cluster CRD patched, but error cleaning environment")
            return f"Error cleaning environment after removing OSDs from Ceph cluster CRD: {e}"
        try:
            self.set_osds_down()
            self.set_osds_out()
            if self.replace_flag:
                self.destroy_osds()
            else:
                self.purge_osds()
        except Exception as e:
            log.exception("OSDs removed from environment, but not able to remove OSDs from Ceph cluster")
            return f"Error removing OSDs from Ceph cluster: {e}"

        return remove_result
```